keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,1017 @@
+
+  The list of such capabilities is likely to change from one Linux version to
+  another. For example, Linux now has [time
+  namespaces](https://man7.org/linux/man-pages/man7/time_namespaces.7.html) and
+  there are ways to make `CAP_SYS_TIME` work inside a user namespace. There are
+  also discussions to make `CAP_MKNOD` work in user namespaces.
+
+- **Sharing Host Namespaces**:
+  There are some limitations in the Linux kernel and in the runtimes that
+  prevent sharing other host namespaces when the host user namespace is not
+  shared.
+  - Mounting `mqueue` (`/dev/mqueue`) is not allowed from a process in a user
+    namespace that does not own the IPC namespace. Pods with `hostIPC=true` and
+    `userNamespaceMode=Pod|Cluster` can fail.
+  - Mounting `procfs` (`/proc`) is not allowed from a process in a user namespace
+    that does not own the PID namespace. Pods with `hostPID=true` and
+    `userNamespaceMode=Pod|Cluster` can fail.
+  - Mounting `sysfs` (`/sys`) is not allowed from a process in a user namespace
+    that does not own the network namespace. Impact: pods with
+    `hostNetwork=true` and `userNamespaceMode=Pod|Cluster` can fail.
+
+  If users specify `userNamespaceMode=Pod|Cluster` and one of these
+  `host{IPC,PID,Network}=true` options, runc will currently fail to start the
+  container. The kubelet does **not** try to prevent that combination of options,
+  in case runc or the kernel make it possible in the future to use that
+  combination.
+
+In order to avoid breaking existing workloads `Host` is the default value of `userNamespaceMode`.
+
+#### Duplicated Snapshots of Container Images
+
+The owners of the files of a container image have to been set accordingly to the
+ID mappings used for that container. For example, if the user 0 in the container
+is mapped to the host user 100000, then the `/root` directory has to be owned by
+user ID 100000 in the host, so it appears to belong to root in the container.
+The current implementation in container runtimes is to recursively perform a
+`chown` operation over the image snapshot when it's pulled. This presents a risk
+as it potentially increases the time and the storage needed to handle the
+container images.
+
+There is not immediate mitigation for it, [we
+talked](https://lists.linuxfoundation.org/pipermail/containers/2020-September/042230.html)
+to the Linux kernel community and [they
+replied](https://lists.linuxfoundation.org/pipermail/containers/2020-September/042230.html)
+they are working on a solution for it. If the Linux kernel provides a solution
+for this problem, that would be something that container runtimes should use. It
+does not impact the kubelet nor the CRI gRPC spec.
+
+Another risk is exausting the disk space on the nodes if pods are repeativily
+started and stopped while using `Pod` mode. Since `Pod` mode is planned for
+phase 2 we haven't considered a mitigation for this case.
```

Comment:
We need a section on impact to image gc and rootfs for all cases.

Reply:
I changed this section a bit based on our meeting and put a new section about garbage collection in the Pod mode -> https://github.com/kinvolk/kubernetes-enhancements/pull/3/commits/6093bac774d4c76f3c417a93e69021d21ee33679

Please let me know if that's what you meant.